src/mem/localalloc.h

```diff
@@ -459,6 +464,14 @@ namespace snmalloc
       //  before init, that maps null to a remote_deallocator that will never
       //  be in thread local state.
 
+#  ifdef __CHERI_PURE_CAPABILITY__
+      // On CHERI platforms, snap the provided pointer to its base, ignoring
+      // any client-provided offset, which may have taken the pointer out of
+      // bounds.  The base, on the other hand, is guaranteed by monotonicity
+      // to be within the bounds originally returned by alloc().
```

Comment:
It can be one-past-the-end though if the capability is zero length; will that be handled correctly?

Reply:
Hrm.  Inasmuch as it's undefined behaviour, yes, but no, that's not what I meant to have happen.  (Such one-past-the-end, zero-length capabilities are also revoked with the wrong object, but since they confer no authority, we've let them slide.  Here, though, this won't work, since we, `snmalloc`, are about to amplify based on the provided address.)

The correct invariant, I think, is that, if the wild cap is indeed a descendant of one of ours, its `base + length - 1` always points within the original allocation.  Sadly, we don't have anything shorter than `CGetLength` `addi -1` `CSetOffset` to get there, I think.  This might have tangential bearing on the discussion around https://github.com/CTSRD-CHERI/cheri-architecture/pull/37 or the old set-offset-relative-to-the-other-end instruction that was proposed.  Until then, here's the three instruction sequence.

That, in turn, brings up the whole "we should actually be reconstructing caps so as to restore any shed permissions" thing.  I made a start at it but I'll push that through more tomorrow.  In the interim, we could keep this hunk as an incomplete improvement or just wait for that.

Reply:
That just shifts the problem to the other end (i.e. `free(cheri_bounds_set(malloc(len), 0))` will give you a zero length cap with base pointing to the first byte and length 0, so that calculation takes you one byte _before_ the allocation).

Reply:
Which makes me wonder if Cornucopia got this right (or if it's even possible to get right...)

Reply:
Cornucopia does actually get this right, inasmuch as those zero-length caps convey no authority and so don't "need" to be revoked per se in order to preserve the invariant that all authority to the address space is removed before the end of the revocation.  It's perhaps a little surprising that some caps might survive that exercise, but go ahead and try to use one, right?

I suppose I'd be happy to reject length-zero caps early in `free` since that can't possibly be what you meant.  And then `base` is always in bounds and we're good, right?

Reply:
(Though I suppose we should write down somewhere that length-zero caps are not to be used as tokens for anything.)

Reply:
> Cornucopia does actually get this right, inasmuch as those zero-length caps convey no authority and so don't "need" to be revoked per se in order to preserve the invariant that all authority to the address space is removed before the end of the revocation. It's perhaps a little surprising that some caps might survive that exercise, but go ahead and try to use one, right?

Just don't use them as integer type tokens. Which is a good idea anyway, since someone else can come and derive a cap to the same end if their allocation happens to be adjacent to yours. Anything in the middle should be totally safe.

> I suppose I'd be happy to reject length-zero caps early in `free` since that can't possibly be what you meant. And then `base` is always in bounds and we're good, right?

You'd need to be careful about `malloc(0)`; if it returns NULL then everything's fine, but if it returns a unique non-NULL value you need to not leak those underlying allocations.

Reply:
`malloc(0)` currently returns a fresh allocation.  I found better compatibility with that.  An old version of Clang checked for `nullptr` return and said "Out of Memory", but was requested a size of 0.

It is fairly easy to change to returning `nullptr`, we could make that a configuration option?

Reply:
I think we're fine as it stands re: `malloc(0)`.  To ensure that we never hit unrepresentable bounds due to capability compression, we always bound to the underlying sizeclass size rather than the requested size.

Reply:
That will fail cheribsdtest. Bounds must be as tight as they can be.

Reply:
If you bound tightly, you have to rebound on free.  For the case where you don't have any UAF Cornucopia like thing.  I will add things to a queue using the internal space of the deallocated object.  So
```C++
  free(malloc(7))
```
will currently fail for CHERI, as I will write to the first 16 bytes of the allocation. To create the `next` and the `signed_prev` pointers.  We are currently make a strong assumption of the first two pointers being accessible.  @nwf you can probably correct me here.  Also, is the min allocation on CHERI 32 bytes?

Reply:
> That will fail cheribsdtest. Bounds must be as tight as they can be.

I don't think that this is required by the CHERI C abstract machine and it is actively harmful as a thing to test because it constrains the implementation space in a way that is surprising to C programmers.

Reply:
Globals and stack allocations are always tightly bounded, it'd be more surprising to CHERI C programmers if malloc were not consistent with that. Yes, it's constraining, no, it's not strictly required for spatial safety, and no, we don't sufficiently specify what the bounds you actually get from malloc are, but when your hardware can represent bounds of 1 it's rather sad for malloc(1) not to have such bounds.

Reply:
Yes, the realloc issue is awkward, but that's a problem with people assuming things that the C standard doesn't quite say is true.

malloc_usable_size, well, people are abusing it. The Linux man-pages project very clearly has:

> The value returned by malloc_usable_size() may be greater than the requested size of the allocation because of alignment and minimum size constraints. Although the excess bytes can be overwritten by the application without ill effects, this is not good programming practice: the number of excess bytes in an allocation depends on the underlying implementation.
> 
> The main use of this function is for debugging and introspection.

I also don't see many uses of it according to codesearch.debin.net (which needs a lot of filtering to avoid picking up all the embedded dlmalloc's and jemalloc's).

Yes, people may be more prone to forgetting bounds aren't always precise, but I don't think "being stricter about X means people forget to be careful about Y therefore we shouldn't be strict about X" is right. Especially when it's already true for globals and stack allocations. Also, if you make it dependent on whatever internal buckets the allocator has then you can end up with more surprise, because code that works on one machine may hit a bounds fault on another just because it happened to have an allocator with a bucket size closer to what you asked for.

The test in question was added in May 2016, so it's nothing new https://github.com/CTSRD-CHERI/cheribsd/commit/beef09dc28b18676a5f40987210a8397877c1853

Reply:
So my concern from chatting with Wes to understand Jessica's comments about `malloc(0)` is that we have to implement `malloc(0)` with its own sizeclass.

Here are my assumptions
* We don't want to store a per-allocation size, it is too costly.
   -  Hence, we can't check that you return the same bounds you requested
* To meet the tight bounds of malloc,  `malloc(0)` has 0 bounds
* malloc(0) is backed by an underlying allocation of 32bytes.

I think from @jrtc27's comments, we could do the following
```
auto p = malloc(32);
auto e = pointer_offset(p, 32);
auto z = restrict_bounds(e, 0);
```
So here `z` looks like a valid allocation of the next object in the slab with size 0.  This could now be used to `free` an unrelated allocation.

This means if we are to implement tight bounds, then we will need to alter the design of snmalloc to have a different slab for 0 sized allocations.  This is doable, and isoalloc does this.  But it is a constraint on the design of snmalloc.

Personally, I prefer giving the larger bounds, and you have to return the full original bounds to `free`.  I read that the argument of `free`:

> must have been returned by a previous call to malloc(), calloc() or realloc().

Personally, I view changing the bounds in the same was as offsetting from the base, not something `free` has to deal with.

That said, I am happy to take a patch that gives 0 its own sizeclass, but it will be of dubious benefit outside of conformance for CHERI.

Reply:
> The test in question was added in May 2016, so it's nothing new CTSRD-CHERI/cheribsd@beef09d

@jrtc27 I took a look at that test but it is a `<` not a `!=`.  So it looks like a larger bounds conforms to me?

Reply:
Interesting, in all these years I'd never noticed that

Reply:
No worries.  This has been a super useful discussion for me.  Understanding the corner cases around amplifying a zero permission are super interesting.